extension/sigv4authextension/extension.go

```diff
@@ -0,0 +1,197 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sigv4authextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/sigv4authextension"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	sigv4 "github.com/aws/aws-sdk-go/aws/signer/v4"
```

Comment:
Just had a random thought - would it make sense to go with aws-sdk-go-v2 for this extension? I don't know if we have a directive to block having both versions of the SDK in the collector, code size wise since it's only the signer / creds loading it should be pretty small. And presumably eventually we do want to migrate components to the more modern SDK and drop v1.

Main reason it came to mind is it's nice that the sign method accepts an `http.Request` itself.

https://github.com/aws/aws-sdk-go-v2/blob/main/aws/signer/v4/v4.go#L269

Tradeoff seems to be needing to compute the payload hash yourself but I guess it's just 3 lines

https://github.com/aws/aws-sdk-go-v2/blob/main/aws/signer/v4/v4_test.go#L51

Think it's worth a thought, it would avoid the copy (ioutil.ReadAll) of a copy (GetBody) of the request body we currently have which is a bummer. Telemetry request payloads tend to be pretty huge, maybe even megabytes, so any additional copy is significant.

Reply:
I find no flaws in your reasoning. :)

Reply:
@Aneurysm9 @anuraaga will begin working on shifting the implementation over to use the aws sdk v2; thinking that we can come back to the concerns about the tests regarding credentials above once we shift the implementation over.

Reply:
Made the migration over to using aws sdk v2; two tests now fail, `TestLoadConfig` and `TestGetCredsFromConfig/success_case_with_role`. From prior discussion, it seems like the first failure is somewhat expected(?) and the second failure is expected(though after further thought, not sure why `getCredsFromConfig()` should [fail when no credentials on CI](https://github.com/open-o11y/opentelemetry-collector-contrib/pull/1302#discussion_r805477104) since it's a function to retrieve credentials). `getCredsFromConfig()`'s logic when a `RoleArn` is provided is being worked on to finish up work on credential-related tests.

Reply:
Resolving conversation after discussion was made on tests and they were removed